### PR TITLE
Cross-bank multi-axis supervision + bis_xbank DAPT substrate

### PR DESCRIPTION
**Pull request by yusufizzetmurat**
## Summary

- Add `--cross-bank-supervision off|stance_masked|weighted` flag to multi-axis text trainer
- `stance_masked` arm: cross-bank rows train on certainty/factor/topic only, stance masked
- `weighted` arm: per-row stance loss scaling via new `MultiTaskLoss.stance_sample_weight` kwarg
- New `bis_xbank` DAPT substrate mixing BIS speeches with 5 cross-bank corpora as NSP pairs
- Register `finbert_fed_adjacent_xbank_dapt` encoder alias + Makefile pretrain target
- Manifest records xbank dataset SHAs alongside the BIS revision

Behaviour note: in mode `off` (default) the gtfintechlab loader now restricts to FOMC only. Prior default silently loaded all six banks without stance protection in the multi-axis trainer — a pre-existing silent bias the new gate closes.

## Test plan

- \`docker compose run --rm backend pytest tests/unit/test_continued_pretraining.py tests/unit/test_label_schemas.py tests/unit/test_cross_bank_supervision_flag.py tests/unit/test_train_text_multi_axis_cross_bank_gate.py -q\`
- 1-seed × 4-fold forecaster smoke with the new encoder once Bundle A.2 (GPU) lands the cache
- DAPT smoke: \`make finbert-fed-adjacent-xbank-dapt-pretrain\` once Bundle A.4 kicks off

Sweep result rows for the wiki §6.6 result-improvement table land on this PR before merge.

### CODE REVIEW

File: backend/app/data/continued_pretraining.py [hunk 1]
```diff
@@ -119,6 +148,78 @@ def _bis_pair_stream(
...
+                        a, b = buffer
+                        if a != b:
+                            pairs.append(
+                                {
+                                    "sequenceA": a,
+                                    "sequenceB": b,
+                                    "next_sentence_label": 0,
+                                }
+                            )
+                        buffer = []
```
Comment (yusufizzetmurat) [inline backend/app/data/continued_pretraining.py]: **Important — odd-count and duplicate-sentence buffer drops.** Two related issues in this loop:

1. The buffer is reset `buffer = []` after every pair, but never flushed at the end of a `(config, split)` iteration. If a split has an odd number of non-empty sentences, the last sentence is silently dropped, with no log or manifest record.
2. When the dedup guard `if a != b:` rejects a pair (consecutive identical sentences), `b` is also discarded by the unconditional `buffer = []`. `b` could have legitimately paired with the next sentence `c`. Across 5 banks × multiple configs/splits, this can silently shrink the xbank NSP slice well below the documented floor.

Fix sketch: on the dedup path, instead of `buffer = []`, set `buffer = [b]` so `b` becomes the new candidate first-of-pair. For the trailing odd sentence, either pair it with a copy of itself (then mask via the dedup guard — consistent), or emit a single-sentence record with a synthetic NSP label, or just log a per-split count of dropped trailing sentences for traceability.

File: backend/app/data/continued_pretraining.py [hunk 2]
```diff
@@ -119,6 +148,78 @@ def _bis_pair_stream(
...
+    """
+    from datasets import (  # type: ignore
+        get_dataset_config_names,
+        get_dataset_split_names,
+        load_dataset,
+    )
+
+    pairs: list[dict[str, Any]] = []
+    for _bank_key, dataset_id in _GTFINTECHLAB_XBANK_DATASETS:
+        revision = _GTFINTECHLAB_XBANK_REVISIONS.get(dataset_id)
```
Comment (yusufizzetmurat) [inline backend/app/data/continued_pretraining.py]: **Important — silent revision drop if `_GTFINTECHLAB_XBANK_REVISIONS` is missing a dataset ID.** `revision = _GTFINTECHLAB_XBANK_REVISIONS.get(dataset_id)` returns `None` for an unknown key with no warning. If a future edit adds a bank to `_GTFINTECHLAB_XBANK_DATASETS` but forgets the revision pin, `load_dataset` is called without `revision=`, fetching whatever the Hub HEAD is — silently breaking the pinned-substrate reproducibility invariant. Fix: raise a `KeyError` (or explicit `SystemExit`) when the revision is missing. The two constants should also share a single source of truth or be enforced equal at module load.

File: backend/app/data/train_text_multi_axis_classifier.py [hunk 3]
```diff
@@ -377,7 +484,10 @@ def _load_gtfintechlab_rows(*, fed_only: bool = False) -> list[_AxisRow]:
     # training side stays aligned with the ingestion side.
     from app.data.ingest_sources import _dataset_revision
 
-    dataset_ids = _GTFINTECHLAB_DATASET_IDS[:1] if fed_only else _GTFINTECHLAB_DATASET_IDS
+    if fed_only or cross_bank_mode == "off":
```
Comment (yusufizzetmurat) [inline backend/app/data/train_text_multi_axis_classifier.py:487]: **Minor — `fed_only=True` silently overrides `cross_bank_mode` and the run log lies.** The condition `if fed_only or cross_bank_mode == 'off':` restricts to FOMC whenever `fed_only` is true, even if `cross_bank_mode='stance_masked'` or `'weighted'`. `main()` already emits a warning (good), but the checkpoint payload at lines 950-955 still records `cross_bank_supervision=weighted` — a provenance lie in the saved artifact. Either error out at flag-parse time when both are set, or in the checkpoint payload record an effective-arm field that reflects what actually happened.

**Review by yusufizzetmurat — COMMENTED** [review]
## /code-review findings (extra-high recall pass)

Three findings already posted as inline comments (xbank iterator drops, revision-pin gap, fed-only override lie). Two more critical/important findings target unchanged code in `_evaluate`, so posting here:

### Critical — `_evaluate` (backend/app/data/train_text_multi_axis_classifier.py:892)

`_evaluate` calls `loss_fn(logits, targets, masks)` without threading `stance_sample_weight`. Train uses `_compute_weighted_total_loss` (per-row weighted CE on stance); eval uses plain unweighted CE. Under `--cross-bank-supervision=weighted`, train and val losses are computed on different surfaces. Best-checkpoint selection by `val_loss` therefore picks the epoch that minimised the unweighted objective, not the weighted training objective.

Fix: extract `stance_sample_weight = batch.get("stance_sample_weight")`, move to device if not None, pass through to `loss_fn(..., stance_sample_weight=...)` — same pattern as the train side.

### Important — `_evaluate` builds a graph per batch (lines 864-912)

`model.eval()` only flips dropout/BN to inference mode; it does not disable autograd. Wrap the forward + axis-loss block in `with torch.no_grad():` (or decorate with `@torch.inference_mode()`) so the autograd graph is not retained across batches. Pre-existing pattern in the file but worth fixing alongside the train/val loss alignment.

### Summary of full pass

5 finder angles + 1 sweep over the ~900 LOC diff. Top findings ranked by severity:

1. Critical — `_evaluate` missing `stance_sample_weight` → train/val loss divergence under `weighted` arm
2. Important — `_evaluate` missing `torch.no_grad()` → graph retention / OOM risk
3. Important — xbank pair iterator drops trailing odd sentence + drops `b` when `a==b` (inline)
4. Important — `_GTFINTECHLAB_XBANK_REVISIONS.get` returns `None` silently (inline)
5. Minor — `fed_only=True` + `cross_bank_mode!='off'` produces a provenance lie in the checkpoint metadata (inline)

Items already verified during the per-bundle code-quality reviews (double-gradient fix, zero-weight edge case, dead membership entry, stub signatures, manifest reproducibility) confirmed correct and not regressed.

**Comment by yusufizzetmurat** [comment]
## Review-fix round (commits ad97e2a / 87cf0a7 / 05e6bdc)

Mapping the 5 /code-review findings to the fix commits now on this PR:

| # | Finding | Fix commit |
|---|---------|-----------|
| 1 | Critical — `_evaluate` ignores `stance_sample_weight` → train/val loss divergence under `weighted` arm | `ad97e2a` |
| 2 | Important — `_evaluate` missing `torch.no_grad()` → graph retention per batch | `ad97e2a` |
| 3 | Important — xbank pair iterator drops `b` on `a==b` + drops trailing odd sentence | `87cf0a7` |
| 4 | Important — `_GTFINTECHLAB_XBANK_REVISIONS.get(...)` silently returns `None` for missing keys | `87cf0a7` |
| 5 | Minor — `fed_only=True` + non-`off` cross_bank_mode → provenance lie in checkpoint | `05e6bdc` |

8 new unit tests cover the fixes (train/val parity on a `[1.0, 0.25]` weighted batch, dedup preserves `b` and doesn't double-count when `b==c`, trailing flush logging, `KeyError` on missing revision, effective_cross_bank_supervision derived field). 61 tests pass in the targeted regression sweep (53 baseline + 8 new).
